File: Flutter/pickup_scheduler/lib/utils/Customer.dart
import 'package:geocoder/geocoder.dart';
import 'package:pickup_scheduler/utils/Date.dart';
import 'dart:math';

class Customer {
  String name;
  String address;
  Duration offset;
  Date startDate;
  double x;
  double y;
  Point point;

  Customer(String c) {
    var data = c.split('::');
    this.name = data[0];
    this.address = data[1];
    this.offset = new Duration(days: int.parse(data[2]));
    this.startDate = new Date(int.parse(data[3]), int.parse(data[4]), int.parse(data[5]));
    calculateCoordinates();
  }

  calculateCoordinates() async {
    var loc = await Geocoder.local.findAddressesFromQuery(address);
    if (loc.length >= 1) {
      Coordinates coordinates = loc.first.coordinates;
      x = coordinates.latitude;
      y = coordinates.longitude;
      print("resolved to $x, $y");
    } else {
      x = 47.4874;
      y = 117.5758;
      print("Could not resolve address to coordinates");
    }
    point = new Point(x, y);
  }

  List<Date> getNextNPickupDates(int n, Date min) {
    List<Date> dates = new List<Date>();
    Date currentDate = new Date.fromDateTime(new DateTime.now());
    Date pickupDate = startDate;
    //Fast-forward to minimum date
    while (pickupDate < min) {
      pickupDate = pickupDate.increment(offset);
    }

    //Add in all the pickup dates between min and today
    while (pickupDate < currentDate) {
      dates.add(pickupDate);
      pickupDate = pickupDate.increment(offset);
    }
    dates.add(pickupDate);

    //We can now work forward n
    for (int i = 0; i < n; i++) {
      pickupDate = pickupDate.increment(offset);
      dates.add(pickupDate);
    }

    return dates;
  }

  List<Date> getAllPickupDates(Date min, Date max) {
    List<Date> dates = getNextNPickupDates(8, min);
    List<Date> filteredDates = new List<Date>();

    for (Date d in dates) {
      if (d >= min && d <= max) filteredDates.add(d);
    }

    return filteredDates;
  }
}

File: Native Android/app/src/main/java/com/ibisrecycling/lars/pickupscheduler/utils/Customer.kt
package com.ibisrecycling.lars.pickupscheduler.utils

import android.content.Context
import android.location.Geocoder
import android.widget.Toast
import com.ibisrecycling.lars.pickupscheduler.utils.DateConversions.Companion.getAsInt
import com.ibisrecycling.lars.pickupscheduler.utils.DateConversions.Companion.getAsString
import java.text.SimpleDateFormat
import java.util.*


class Customer(
		val name: String,
		val address: String,
		val subscriptionType: Int,
		val startDate: Date,
		private val context: Context) {

	private val frequency: Long

	override fun toString(): String {
		return "$name::$address::${frequency / daysToMillis}::${getAsString(startDate)}"
	}

	init {
		frequency = subscriptionType * daysToMillis
	}

	companion object {
		private val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.US)
		const val daysToMillis: Long = 86400000
	}

	constructor(s: List<String>, context: Context) : this(
			s[0],
			s[1],//TODO: Addresses contain commas, deal with it
			s[2].toInt(),
			dateFormat.parse(s[3]),
			context)

	constructor(s: String, context: Context) : this(
			s.split("::"),
			context)

	val location: Point by lazy {
		val address = Geocoder(context).getFromLocationName(address, 2)
		if (address != null) {
			val loc = address[0]
			Point(loc.latitude, loc.longitude)
		} else {
			Toast.makeText(context, "Invalid Address $address", Toast.LENGTH_LONG).show()
			Point(0.0, 0.0)
		}
	}
	val x: Double by lazy {
		location.x
	}
	val y: Double by lazy {
		location.y
	}

	fun getNextNPickupDates(n: Int): ArrayList<Date> {
		val dates = ArrayList<Date>()
		val currentDate = Date()
		var pickupDate = startDate

		//Fast-forward the start date to the present date
		while (getAsInt(pickupDate) < getAsInt(currentDate))
			pickupDate = Date(pickupDate.time + frequency)
		dates.add(pickupDate)

		//We can now work forward n - 1 times
		repeat(n - 1, {
			pickupDate = Date(pickupDate.time + frequency)
			dates.add(pickupDate)
		})

		return dates
	}

	fun getNextPickupDate(): Date {
		return getNextNPickupDates(1)[0]
	}
}